### PR TITLE
helm: possibility to control creation of GatewayClass

### LINKED ISSUE #33239: cilium does not create Gateway class
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

I have installed the v1.0.0 Gateway API CRDs and after that performed an install of Cilium using the following:

```bash
helm template \
cilium \
cilium/cilium \
--version 1.16.0-pre.3 \
--namespace kube-system \
-f cilium-values.yml \
> talos/cilium-manifest.yml
```

the inputs file is:

```yaml
ipam:
  mode: kubernetes
kubeProxyReplacement: true
securityContext:
  capabilities:
    ciliumAgent:
      [
        CHOWN,
        KILL,
        NET_ADMIN,
        NET_RAW,
        IPC_LOCK,
        SYS_ADMIN,
        SYS_RESOURCE,
        DAC_OVERRIDE,
        FOWNER,
        SETGID,
        SETUID,
      ]
    cleanCiliumState: [NET_ADMIN, SYS_ADMIN, SYS_RESOURCE]
cgroup:
  autoMount:
    enabled: false
  hostRoot: /sys/fs/cgroup
k8sServiceHost: localhost
k8sServicePort: 7445
gatewayAPI:
  enabled: true
  hostNetwork:
    enabled: true
```

Then I proceed with installing the gateway:

```
apiVersion: gateway.networking.k8s.io/v1
kind: Gateway
metadata:
  name: my-gateway
  namespace: test
spec:
  gatewayClassName: cilium
  listeners:
    - protocol: HTTP
      port: 8080
      name: web-gw
      allowedRoutes:
        namespaces:
          from: Same
```

But then `cilium-operator` complains that no `cilium` gateway class is available:

```
2024-06-18 16:57:07.650Z cilium-operator-c85775f8d-dcn5n time="2024-06-18T16:57:07Z" level=error msg="Unable to get GatewayClass" controller=gateway error="GatewayClass.gateway.networking.k8s.io \"cilium\" not found" resource=my-gateway subsys=gateway-controller
```

I wonder, how do I identify why the operator did not create the gateway class?

### Cilium Version

1.16.0-pre3

### Kernel Version

6.6.32-talos

### Kubernetes Version

1.30.1

### Regression

_No response_

### Sysdump

_No response_

### Relevant log output

_No response_

### Anything else?

_No response_

### Cilium Users Document

- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by balous**
Fixes: #33239

This PR adds possibility to Cilium helm chart to controll installation of `GatewayClass` by adding `gatewayAPI.gwClass.create: <true|false|auto>` option to helm chart's values. `auto` (the default) is the original behavior where the installation was controlled by presence of GatewayAPI in the target cluster. When `true` is selected and GatewayAPI is not present in the target cluster, Cilum installation fails (the same way it fails when prometheus-operator CRDs are not present and monitoring is enabled).

```release-note
helm: Add possibility to control creation of GatewayClass
```

### CODE REVIEW

File: install/kubernetes/cilium/values.yaml.tmpl [hunk 1]
```diff
@@ -853,6 +853,13 @@ gatewayAPI:
   # -- Control how traffic from external sources is routed to the LoadBalancer Kubernetes Service for all Cilium GatewayAPI Gateway instances. Valid values are "Cluster" and "Local".
   # ref: https://kubernetes.io/docs/reference/networking/virtual-ips/#external-traffic-policy
   externalTrafficPolicy: Cluster
+
+  gwClass:
```
Comment (sayboras) [inline install/kubernetes/cilium/values.yaml.tmpl]: probably just use gatewayClass in full form to align with other existing field (e.g. gatewayAPI)

Comment (balous) [reply install/kubernetes/cilium/values.yaml.tmpl]: Changed.

File: install/kubernetes/cilium/values.yaml.tmpl [hunk 2]
```diff
@@ -854,6 +854,13 @@ gatewayAPI:
   # Note that this value will be ignored when `hostNetwork.enabled == true`.
   # ref: https://kubernetes.io/docs/reference/networking/virtual-ips/#external-traffic-policy
   externalTrafficPolicy: Cluster
+
+  gatewayClass:
+    # -- Enable creation of GatewayClass resource
+    # default value is 'auto' which decides according to presence of gateway.networking.k8s.io/v1/GatewayClass in the cluster
+    # other possible values are 'true' and 'false'
```
Comment (youngnick) [inline install/kubernetes/cilium/values.yaml.tmpl]:
```suggestion
    # The default value is 'auto' which decides according to presence of gateway.networking.k8s.io/v1/GatewayClass in the cluster.
    # Other possible values are 'true' and 'false', which will either always or never create the GatewayClass, respectively.
```
Small edit for grammar here, which should make this render out more nicely in the other files.

**Review by youngnick — COMMENTED** [review]
One small change for grammar and this LGTM. Thanks @balous!

**Comment by youngnick** [comment]
I think this will also need a `make -C install/kubernetes ciilum/values.yaml docs` to update all the generated files.

**Comment by balous** [comment]
> I think this will also need a `make -C install/kubernetes ciilum/values.yaml docs` to update all the generated files.

Done.

**Review by qmonnet — COMMENTED** [review]
Looks good as far as I'm concerned, thanks!

**Review by qmonnet — APPROVED** [review]

**Review by sayboras — APPROVED** [review]

**Pull request merged**